code/components/jomjol_controlcamera/ClassControllCamera.cpp

```diff
@@ -219,144 +244,273 @@ void CCamera::SetCameraFrequency(int _frequency)
-        */
-
-        //s->set_reg(s, 0x7C, 0xFF, 2); // Optional feature - hue setting: Select byte 2 in register 0x7C to set hue value
-        //s->set_reg(s, 0x7D, 0xFF, 0); // Optional feature - hue setting: Hue value 0 - 255
-        s->set_reg(s, 0xFF, 0x01, 0); // Select DSP bank
-        s->set_reg(s, 0x7C, 0xFF, 0); // Select byte 0 in register 0x7C
-        s->set_reg(s, 0x7D, 7, 7); // Set bit 0, 1, 2 in register 0x7D to enable saturation, contrast, brightness and hue control
+    if (s != NULL) {
+        // auto exposure controls
+        s->set_exposure_ctrl(s, 1); // Enable auto exposure control
+        s->set_gainceiling(s, GAINCEILING_2X); // GAINCEILING_2X 4X 8X 16X 32X 64X 128X -> Limit gain
+        s->set_ae_level(s, std:: min(2, std::max(-2, _autoExposureLevel))); // -2 .. 2 -> Adapt auto exposure control level
+        s->set_aec2(s, _aec2Algo ? 1 : 0); // Switch to alternative auto exposure control alogrithm
+
+        // post processing
+        s->set_brightness(s, std::min(2, std::max(-2, _brightness)));   // -2 .. 2
+        s->set_contrast(s, std::min(2, std::max(-2, _contrast)));       // -2 .. 2
+        s->set_saturation(s, std::min(2, std::max(-2, _saturation)));   // -2 .. 2
+        s->set_sharpness(s, 0); // Not officially supported, use own implementation
+
+        camera_sensor_info_t *sensor_info = esp_camera_sensor_get_info(&(s->id));
+        if (sensor_info != NULL) {
+            if (sensor_info->model == CAMERA_OV2640) {
+                // Sharpness implementation (not officially supported)
+                if (autoSharpness) {
+                    s->set_sharpness(s, 0);
+                    ov2640_enable_auto_sharpness(s);
+                } 
+                else {
+                    ov2640_set_sharpness(s, std::min(3, std::max(-3, _sharpness))); // -3 .. 3
+                }
+
+                /* Workaround - bug in cam library - enable bits are set without using bitwise OR logic -> only latest enable setting is used */
+                /* Library version: https://github.com/espressif/esp32-camera/commit/5c8349f4cf169c8a61283e0da9b8cff10994d3f3 */
+                /* Reference: https://esp32.com/viewtopic.php?f=19&t=14376#p93178 */
+                //s->set_reg(s, 0x7C, 0xFF, 2); // Optional feature - hue setting: Select byte 2 in register 0x7C to set hue value
+                //s->set_reg(s, 0x7D, 0xFF, 0); // Optional feature - hue setting: Hue value 0 - 255
+
+                int indirectReg0 = 0x07; // Set bit 0, 1, 2 to enable saturation, contrast, brightness and hue control
+                
+                // Bitwise OR grayscale and/or negative
+                if (_grayscale && _negative) {
+                    indirectReg0 |= 0x58;
+                }
+                else if (_grayscale) {
+                    indirectReg0 |= 0x18;
+                }
+                else if (_negative) {
+                     indirectReg0 |= 0x40;
+                }
```

Comment:
I have 2 comments on this part:
1. These if else statements are redundant. You are setting bits of an effect is enabled, it's cleaner to write it this way:
```
if (_grayscale) {
    indirectReg0 |= 0x18;
}
if (_negative) {
    indirectReg0 |= 0x40;
}
```
When both grayscale and negative are enabled, you end up with 0x58 anyway.

2. The camera will perform one effect only even if both grayscale and negative are enabled. It will give preference to grayscale effect and ignore negative. That is why I chose to do negative effect on the ESP and web browser instead.

Reply:
@jasaw: Thanks for your review :-)

Indeed, it seems that both special effects together seems not working, at least it's hardly to identify. So I decided to switch to 'exlusive or' functionality, either grayscale or negative. I assume grayscale and negative together would be rarly needed.

Why so? I try to keep as much of basic image manipulation on camera side to have those settings also available for all possible image consumers, e.g. live stream or REST API. Therefore I even enabled mirror functionality on camera side to get rid of this in alignment state.

Your approach is also a good solution if both are needed together. Thanks again for your research. It's quite difficult to understand those badly documented camera registers :-)

One question regarding zoom functionality:
Have you seen any possiblitly to modify zoom factor? I realized that there is quite high likelyhood one zoom mode is to high and other to low. I've just read something about a zoom factor in the manual, but haven't further thought about it...

Reply:
@Slider0007 You're welcome. :-)

The reason why I wanted grayscale and negative effects together is because applying grayscale first then negative removes the coloured background on some of my meter digits, making all the digits look consistent, then inverts the colour to give me black digits which is easier to read. For my set up, to improve the AI inference, I'm in the process of retraining my digit recognition model with grayscale only training set to "teach" the AI that colour is not important.

I have thought about the other camera consumers but don't really have a clean way of supporting both grayscale and negative effects together, except for implementing negative effect on the html side as well. Yes the ov2640 datasheet is missing documentation on the indirect registers, otherwise we might be able to figure out how to get the camera to apply both grayscale and negative effects together.

Regarding the zoom (or called windowing (with sub-sampling) on the ov2640 datasheet) functionality, the 2 levels are the officially supported levels. I haven't seen anywhere that talks about other zoom levels, might have to contact OmniVision directly. Note that this zoom level or windowing sub-sampling is different from non-standard output resolution. Setting the non-standard output resolution is definitely possible, as long as we conform to the restrictions like width and height must be divisible by 8, etc.

Reply:
@jasaw: Your approach sounds reasonable.

As far I figured out, it should be possible to apply grayscale + negative at the same time. It's at least mentioned in [Software Application Notes](https://www.dragonwake.com/download/camera/ov2640/OV2640%20Camera%20Module%20Software%20Application%20Notes.pdf). But 0x58 seems equal output to negative only. I assume only the vendor will probably know...

![image](https://github.com/Slider0007/AI-on-the-edge-device/assets/115730895/a24f671d-f063-4ff5-a00a-01d409cd8618)

Thanks for sharing your knowledge in terms of zoom levels.

Reply:
@Slider0007 Thank you for sharing the software application note. I tried setting "remove UV" bit together with negative effect, but the negative effect always takes precedence and remove UV is ignored as well, the same behaviour as setting 0x58. It might be a firmware bug in the ov2640 making grayscale and negative together impossible on the camera.

I would still do the negative on the ESP and web browser HTML side, even for the live stream. Just wrap the live stream in `<img src=getDomainname() + '/stream?flashlight=true' style='filter: invert(1);'>`

Reply:
@jasaw: Short final feedback to this topic. I've done some more trail and error, but grayscale+negative is not working like described in documentation. Now I descided to handle this situation almost identical than your implemenation. I do negative also on MCU but only if grayscale + negative is selected. A single negative or grayscale will still be performed on camera. I'm going to accept this drawback because I was not able to identify any smart idea to have identical presentation in livestream as well as REST API for this special situation. All other combination are fully supported though.
Further I added manual exposure and gain control parameter to WebUI. It seems that in static environment like we have in this meter application without any exteranl light distortion image seems more stable if automatic exposure and gain control is not enabled. But this could be also just a personal interpretaion. I will see... :-)

Thanks again for open up this topic. It was quite interessing to jump into this topic...

Reply:
>I do negative also on MCU but only if grayscale + negative is selected. A single negative or grayscale will still be performed on camera.

This is a great idea. Offload as much as possible to the camera.

>Further I added manual exposure and gain control parameter to WebUI.

I have been playing with auto and manual exposure and gain control on my local dev environment to improve the image quality and this is what I found:
1. If you have a fully enclosed ESP32cam housing on the meter (where lighting is controlled), manual exposure and gain control is fantastic.
2. If your fully enclosed ESP32cam housing is not thick enough and direct sunlight can shine through the 3D printed plastic, it's best to have automatic exposure and gain control.
3. If your ESP32cam is half a meter away from the meter, there is no way of controlling the light source (except at night), even automatic exposure and gain control is not good enough. I have this set up for my gas meter reader because of IECEx regulatory requirements (ATEX for the EU).

I'm playing with a few ideas on how to improve the image quality of scenario 3:
1. Use custom output resolution to only cover the ROI areas. The automatic exposure and gain control appears to be operating on the windowed section of image only. More experiments are needed to confirm this theory.
2. Take multiple photos, iterating through different manual exposure level until ROI areas are not over-exposed or under-exposed. The starting manual exposure level may be determined by the ambient brightness.
To read the ambient brightness:
```
s->set_reg(s,0xff,0xff,0x01);
int brightness = s->get_reg(s,0x2f,0xff);
```

This is a very interesting project and I'm happy to be able to contribute. :-)